File: packages/i18nify-js/src/modules/i18n/flattenKeys.ts
import { withErrorBoundary } from '../../common/errorBoundary';
import { DEFAULT_DELIMITER } from './constants';
import type { FlattenOptions, JsonObject } from './types';

function _flatten(
  obj: JsonObject,
  delimiter: string,
  prefix: string,
  result: JsonObject,
): void {
  for (const key of Object.keys(obj)) {
    const fullKey = prefix ? `${prefix}${delimiter}${key}` : key;
    const value = obj[key];
    if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
      _flatten(value as JsonObject, delimiter, fullKey, result);
    } else {
      result[fullKey] = value;
    }
  }
}

const flattenKeys = (obj: JsonObject, options?: FlattenOptions): JsonObject => {
  if (obj === null || typeof obj !== 'object' || Array.isArray(obj)) {
    throw new Error('flattenKeys: input must be a plain object.');
  }
  const delimiter = options?.delimiter ?? DEFAULT_DELIMITER;
  if (!delimiter) {
    throw new Error('flattenKeys: delimiter must be a non-empty string.');
  }
  const result: JsonObject = {};
  _flatten(obj, delimiter, '', result);
  return result;
};

export default withErrorBoundary<typeof flattenKeys>(flattenKeys);

File: packages/i18nify-go/modules/i18n/flatten_keys.go
package i18n

import (
	"fmt"
)

const defaultDelimiter = "."

// FlattenOptions configures FlattenKeys behaviour.
type FlattenOptions struct {
	// Delimiter separates nested key segments. Defaults to ".".
	Delimiter string
}

// FlattenKeys recursively flattens a nested map[string]any into a single-level
// map using delimiter-joined keys. Arrays are treated as leaf values and are
// not recursed into.
func FlattenKeys(obj map[string]any, opts *FlattenOptions) (map[string]any, error) {
	if obj == nil {
		return nil, fmt.Errorf("i18n: FlattenKeys: input must be a non-nil map")
	}
	delimiter := defaultDelimiter
	if opts != nil && opts.Delimiter != "" {
		delimiter = opts.Delimiter
	}
	result := make(map[string]any)
	flattenRecursive(obj, delimiter, "", result)
	return result, nil
}

func flattenRecursive(obj map[string]any, delimiter, prefix string, result map[string]any) {
	for k, v := range obj {
		fullKey := k
		if prefix != "" {
			fullKey = prefix + delimiter + k
		}
		if nested, ok := v.(map[string]any); ok {
			flattenRecursive(nested, delimiter, fullKey, result)
		} else {
			result[fullKey] = v
		}
	}
}
